File: math/202.py
class Solution:
    def isHappy(self, n):
        """
        :type n: int
        :rtype: bool
        """
        sums_set = set()
        
        while True:
            n = sum([int(i) ** 2 for i in str(n)])
            
            if n == 1:
                return True
            elif n in sums_set:
                return False
            else:
                sums_set.add(n)

           

File: hashtable/202.java
import java.util.Map;
import java.util.HashMap;

class Solution {
    public boolean isHappy(int n) {
        Map<Integer, Boolean> map = new HashMap<>();
        while (true) {
            n = this.getSum(n);
            if (n == 1) {
                return true;
            }
            if (map.containsKey(n)) {
                return false;
            } else {
                map.put(n, true);
            }
        }
    }
    
    private int getSum(int n) {
        int ret = 0;
        while (n > 0) {
            int i = n % 10;
            n = n / 10;
            ret += i * i;
        }
        return ret;
    }
}
